modules/validations/invalid_format.js

```diff
@@ -8,6 +8,51 @@ import * as edtf from 'edtf';
 export function validationFormatting() {
     var type = 'invalid_format';
 
+    /**
+     * Converts a date in OSM approximate date format to EDTF.
+     */
+    function edtfFromOSMDate(osm) {
+        // https://wiki.openstreetmap.org/wiki/Key:start_date#Approximations
+        // https://wiki.openstreetmap.org/wiki/Special:Diff/510218/557626#Proposal_for_date_formatting
+        let [match, year, bc] = osm.match(/^~(\d+)( BC)?$/) || [];
+        if (match) {
+            if (bc) year = -parseInt(year, 10) + 1;
+            return `${year}~`;
+        }
+
+        let circa, decade;
+        [match, circa, decade, bc] = osm.match(/^(~)?(\d+)0s( BC)?$/) || [];
+        if (match) {
+            if (!circa) circa = '';
+            if (!bc) return `${decade}X${circa}`;
+            let startYear = -parseInt(decade, 10) * 10 - 8;
+            let endYear = -parseInt(decade, 10) * 10 + 1;
+            return `${startYear}${circa}/${endYear}${circa}`;
+        }
+
+        let century;
+        [match, circa, century, bc] = osm.match(/^(~)?C(\d+)( BC)?$/) || [];
+        if (match) {
+            if (!circa) circa = '';
+            if (!bc) return `${parseInt(century, 10) - 1}XX${circa}`;
+            let startYear = (parseInt(century, 10) - 1) * -100 - 98;
+            let endYear = (parseInt(century, 10) - 1) * -100 + 1;
+            return `${startYear}${circa}/${endYear}${circa}`;
+        }
+
+        let end;
+        [match, end] = osm.match(/^before (\d{4}(?:-\d\d)?(?:-\d\d)?)$/) || [];
+        if (match) {
+            return `[..${end}]`;
```

Comment:
Based on https://github.com/OpenHistoricalMap/issues/issues/547#issuecomment-1567475475, I’ve been going with the set notation in EDTF level 2 instead of the extended interval notation in EDTF level 1. However, EDTF.js rejects qualifiers in sets, which prevents us from translating early/mid/late modifiers with approximation qualifiers. Academic systems such as the [Bayou City DAMS](https://uhlibraries-digital.github.io/bcdams-map/guidelines/date) apparently rely on approximation in intervals for this purpose.